### PR TITLE
Fix performance problems with `branch.delete`

### LINKED ISSUE #4072: Why is branch.delete slow?
When I run branch.delete it regularly takes 10+ seconds. What takes it so long? I would expect it to essentially be the deletion of a name -> hash mapping. Is it doing a vacuum/gc of newly unreachable state?

I think that it should either be made faster or should provide a message about what it's doing.

### PR DESCRIPTION
**Pull request by tstat**
Fixes #4072

## Overview

See [this comment](https://github.com/unisonweb/unison/issues/4072#issuecomment-1650562835) for an overview.

## Interesting/controversial decisions

There was back-and-forth discussion on slack as to whether we should be consing empty `Branch`es onto child causals that exist in the base `Branch` but not the head `Branch`, or if we should instead just leave them deleted in the head. 

We initially planned to leave them deleted only for project branch deletions, but then decided to leave them deleted everywhere.

## Test coverage

Tested manually, and tested by everything that calls some `updateAt` variant.

### CODE REVIEW

**Comment by aryairani** [comment]
I think any weirdness this change may cause is likely best fixed in the merge algorithm, which we're currently reworking anyway.

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 1]
```diff
@@ -747,18 +747,18 @@ consBranchSnapshot headBranch baseBranch =
...
-    combineChildren :: These (Branch m) (Branch m) -> Branch m
+    combineChildren :: These (Branch m) (Branch m) -> Maybe (Branch m)
     combineChildren = \case
       -- If we have a matching child in both base and head, squash the child head onto the
       -- child base recursively.
-      (These base head) -> head `consBranchSnapshot` base
-      -- This child has been deleted, recursively replace children with an empty branch.
-      (This base) -> empty `consBranchSnapshot` base
+      (These base head) -> Just (head `consBranchSnapshot` base)
+      -- This child has been deleted, let it be
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:755]: rest in peace

**Review by aryairani — APPROVED** [review]

**Review by ChrisPenner — APPROVED** [review]
Nice!
🤞🏼🤞🏼 🤞🏼

**Comment by ceedubs** [comment]
Thanks! branch deletion went from ~30 seconds to ~40ms for me!

**Pull request merged**